### PR TITLE
Fix Zeitwerk 2.6.4 compatibility

### PR DESCRIPTION
**Pull request by jeffmeyers**
## What are you trying to accomplish?

Zeitwerk 2.6.4 was released today and included [this commit](https://github.com/fxn/zeitwerk/commit/af1d6ac6ec687aacb361bcdd71edf304b7ce6ad8), which removes the `root_dirs` attribute and instead uses `__roots`.

Running `bin/rake test` with Zeitwerk updated to 2.6.4 yielded errors:

```
Error:
Packwerk::ApplicationValidatorTest#test_validity:
NoMethodError: undefined method `root_dirs' for #<Zeitwerk::Loader:0x00007f7d5fbcc788 @inflector=...
```

## What approach did you choose and why?

I've updated the `root_dirs` call to instead use `__roots`, which is what the commit referenced above does internally. This is a breaking change, but it looks like there's a [major release pending](https://github.com/Shopify/packwerk/pull/248) and this may be a good candidate for inclusion.

## What should reviewers focus on?

Let me know if this should instead check if the loader responds to certain methods to ensure backwards compatibility.

## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Non-breaking change (a change that doesn't alter functionality - i.e., code refactor, configs, etc.)

### Additional Release Notes

None

## Checklist

- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes (existing specs cover this change).
- [ ] It is safe to rollback this change.

### CODE REVIEW

**Review by gmcgibbon — COMMENTED** [review]
Thanks for the PR.

- Please sign the CLA for your change to be mergable
- I think it would be better if we introduced logic to make this work in both cases (if Zeitwerk is version X, do this, otherwise, do that). This way we can get the patch out sooner.

File: test/support/stub_const.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# typed: true
+# frozen_string_literal: true
+
+module StubConst
+  def with_stubbed_const(klass, const, val)
```
Comment (gmcgibbon) [inline test/support/stub_const.rb:5]: You're going to need a shim for minitest/test to give Sorbet a hint that this method exists.

`in sorbet/rbi/shims/minitest/test.rb`
```
# typed: true

class Minitest::Test
  include StubConst
end
```

**Review by gmcgibbon — COMMENTED** [review]
LGTM after comments are resolved and rubocop + sorbet are passing. The CLA still needs to be signed as well according to the check. If you rebase, the other CI issue is fixed on main.

File: test/support/stub_const.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+    original_value = klass.const_get(const)
+    klass.const_set(const, val)
+    yield
+    klass.const_set(const, original_value)
```
[the remark this answers is not in the document]
Comment (jeffmeyers) [reply test/support/stub_const.rb:10]: Great catch 👍

File: lib/packwerk/loader.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+    def self.autoloaders
+      Rails.autoloaders.map do |loader|
+        new(loader)
+      end
+    end
+
+    def root_dirs
+      return super if Zeitwerk::VERSION < ROOT_DIRS_DEPRECATION_VERSION
+      __roots
```
[the remark this answers is not in the document]
Comment (jeffmeyers) [reply lib/packwerk/loader.rb:19]: Tapioca is segfaulting for me at the moment, trying to resolve.

**Comment by gmcgibbon** [comment]
Thanks!

**Comment by fxn** [comment]
The whole point of that release is to hide the private interface! If you just switch to `__roots` which has two leading underscores screaming PRIVATE! we are at square one.

The public interface for listing root directories is `dirs`.

**Comment by fxn** [comment]
If you look at the patch, you'll see `root_dirs` was tagged as `@private`, and it was undocumented. `root_dirs` has been always private interface because Zeitwerk does not expose the collections where it stores state.

The new double leading underscore variant is there for internal use in the gem, not for external usage!

**Comment by alexevanczuk** [comment]
Thank you for this additional context @fxn. It looks like `__roots` and `dirs` return different values:

```
[7] pry(#<Packwerk::Loader>)> __getobj__.__roots
=> {"/Users/alex.evanczuk/workspace/packwerk/tmp/fixtures/test_check_package_manifests_for_privacy_returns_an_error_for_constants_without______prefix-2022_11_03_09_58_48/components/sales/app/models"=>
  Object}
[8] pry(#<Packwerk::Loader>)> __getobj__.dirs
=> ["/Users/alex.evanczuk/workspace/packwerk/tmp/fixtures/test_check_package_manifests_for_privacy_returns_an_error_for_constants_without______prefix-2022_11_03_09_58_48/components/sales/app/models"]
```

I believe we depend on knowing these default namespaces in `constant_resolver` to know what files define what fully qualified constant, taking default namespaces into account. [That code is here.](https://github.com/Shopify/constant_resolver/blob/main/lib/constant_resolver.rb#L79)

Is there another public API that we can use to get both the root/autoloaded directories and their default namespaces?

**Comment by fxn** [comment]
Note also that I went an extra mile and the `__roots` method is not even explictly defined! It is defined as a side-effect of

```
internal :roots
```

which is screaming INTERNAL and does not even hint at the existence of `__roots`.

**Comment by alexevanczuk** [comment]
I understand that there is already a lot of great dev feedback suggesting not to use something (lack of documentation, @private, double underscore).  Mostly I was just thinking if there was something at runtime that could say very clearly something is private and force someone to use `send` (which also fires off alarms).

Something I like to do in gems is make *everything* private by putting all private stuff in a module called `Private` that I mark with `private_constant :Private`. What's nice about this approach though is the developer has to very explicitly type out that they are using Private (by referencing the word and by explicitly bypassing it), which is very strong feedback.

However, this requires internally separating out the parts of a class which are private to external users but "public" internally, which I think is what you are saying is the main blocker for this, which is understandable as that requires changes to the internal structure that may not be desirable in your situation. 

Thanks for sharing context on that decision!

**Comment by exterm** [comment]
> a way to mark a method as private outside of the context of a module but not inside.

Isn't that exactly what `private_constant` does?

```ruby
module A
  module B
    def self.hello_world
      puts "hello world"
    end
  end
  private_constant :B
  def self.hello_world
    B.hello_world
  end
end
```

```
irb(main):015:0> A.hello_world
hello world
=> nil                                             
irb(main):016:0> A::B.hello_world
(irb):16:in `<main>': private constant A::B referenced (NameError)
```

**Pull request merged**